library/Translate/test/Context/Environment.hs

```diff
@@ -0,0 +1,250 @@
+    "Types properly add all to to the closure"
+    [ T.testCase "top level type" $ do
+        [print] <-
+          capture "type foo a b c = Cons (print-closure 3)" trigger
+        Env.keys print T.@=? Set.fromList ["a", "b", "c"],
+      --
+      T.testCase "let-type properly adds constructors" $ do
+        [inside, body] <-
+          capture
+            "let f = \
+            \ let type foo a b c = \
+            \  | Cons (print-closure 3) \
+            \  | Nil \
+            \ in print-closure 4"
+            trigger
+        Env.keys inside T.@=? constructors
+        Env.keys body T.@=? constructors
+    ]
+  where
+    constructors =
+      Set.fromList ["a", "b", "c", "Nil", "Cons"]
+    trigger =
+      (== "print-closure")
+
+caseTest :: T.TestTree
+caseTest =
+  T.testGroup
+    "Case binder"
+    [ T.testCase "case properly adds bound arguments" $ do
+        [cons, nil] <-
+          capture
+            "let f = \
+            \ case foo of \
+            \  | Cons a b -> (print-closure 3) \
+            \  | Nil -> (print-closure 3)"
+            trigger
+        Env.keys cons T.@=? Set.fromList ["a", "b"]
+        Env.keys nil T.@=? Set.fromList []
+    ]
+  where
+    trigger =
+      (== "print-closure")
+
+lambdaTest :: T.TestTree
+lambdaTest =
+  T.testGroup
+    "λ binder"
+    [ T.testCase "λ properly adds binders" $ do
+        [lamb] <-
+          capture
+            "let f = \\(Cons a b) -> (print-closure 3)"
+            trigger
+        Env.keys lamb T.@=? Set.fromList ["a", "b"]
+    ]
+  where
+    trigger =
+      (== "print-closure")
+
+declaimTest :: T.TestTree
+declaimTest =
+  T.testGroup
+    "declaim binder"
+    [ T.testCase "declaim properly adds info" $ do
+        [lamb] <-
+          capture
+            "let f = declare infixl (+) 7 in print-closure 2"
+            trigger
+        -- we could check for info, but this is sufficient for it
+        -- properly working
+        Env.keys lamb T.@=? Set.fromList ["+"]
+    ]
+  where
+    trigger =
+      (== "print-closure")
+
+openTest :: T.TestTree
+openTest =
+  T.testGroup
+    "open Tests"
+    [ T.testCase "open properly adds symbols" $ do
+        Right (ctx, _) <-
+          Contextualize.contextifyS
+            ( ("Foo", parseDesugarSexp "let f = open A in print-closure 2")
+                :| [("A", parseDesugarSexp "let bar = 3")]
+            )
+        let (_, Cap _ [Env.Closure capture]) =
+              runCtx (Env.passContextSingle ctx trigger recordClosure) emptyClosure
+        Map.toList capture T.@=? [("bar", Env.Info Nothing [] (Just "A"))]
+    ]
+  where
+    trigger =
+      (== "print-closure")
+
+capture :: ByteString -> (NameSymbol.T -> Bool) -> IO [Env.Closure']
+capture str trigger = do
+  Right (ctx, _) <-
+    contextualizeFoo str
+  let (_, Cap _ capture) =
+        runCtx (Env.passContextSingle ctx trigger recordClosure) emptyClosure
+  pure capture
+
+-- Right (ctx,_) <- contextualizeFoo "let f = 3"
```

Comment:
What are these comments for?

Reply:
Just for quickly copy and pasting something into the repl